### ISSUE TITLE
Eliminate redundant use of typename throughout library

### ISSUE THREAD

**Issue by AlisdairM** [cplusplus/draft#3637]
Following the adoption of http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0634r3.html many uses of ‘typename’ throughout the library are now redundant.  They should be eliminated to provide a cleaner modern presentation of the standard library, much as when we switched from typedef to using.

Such a wide-reaching change is likely a policy decision to be approved by LWG, but application is more likely an editorial issue (requiring careful review) to avoid overloading precious LWG time.

Suggest this should be an issue for C++23 to avoid disrupting the imminent progress of C++20, and allowing adequate time for reviews.  May consider updating one clause at a time, to ease review load.

**Comment by jensmaurer** [comment]
Please submit an LWG issue for the policy decision. It would also be helpful to identify patterns of "typename" use that are deemed redundant. (Possibly, not all technically-redundant "typename"s should be removed, for clarity of exposition.)

**Comment by AlisdairM** [comment]
I will try to put a paper into the post-Prague mailing, collecting the different patterns of typename usage that can safely be simplified, and those that cannot.  Hopefully much of the discussion can occur online, prior to the Varna meeting, and without distracting from Prague and C++20.

**Comment by AlisdairM** [comment]
Missed post-prague, but expect a paper with detailed examples for reference for future library wording too in plenty of time for the pre-Varna mailing.

**Comment by AlisdairM** [comment]
The paper was in the April mailing: http://wg21.link/P2150
Using wg21.link reference so we will get a 'live' link to any updates following LWG feedback.

**Comment by zhihaoy** [comment]
All `typename` in "`= typename`" can be removed.

**Comment by jensmaurer** [comment]
Milestone adjusted.

**Comment by AlisdairM** [comment]
@jwakely Is C++26 still a reasonable target for this, or should we preemptively defer to C++29?

I do not expect to have the cycles to update P2150 until after we send the C++26 CD for ballot.
If we review R0 as written, it is likely out of date on some specifics, but may be good enough to adopt wording policy changes that I would find time to apply in time as editorial PRs in time for the CD.

**Comment by jwakely** [comment]
I think we can aim for C++26.

**Comment by AlisdairM** [comment]
Should we schedule P2150R0, as is, for LWG discussion, possibly during a telecon rather than taking precious face-to-face time?  Or do you need an update from me first (that I will be hard pressed to produce in time for C++26).

**Comment by frederick-vs-ja** [comment]
Can we continue on this? I think I've completed a patch for the current working draft.

**Comment by jensmaurer** [comment]
We're still waiting for LWG buy-in to the kinds of situations where we want to editorially remove "typename".  I've updated  cplusplus/papers#862 accordingly.

**Comment by frederick-vs-ja** [comment]
IMO we can discuss this in small parts, each of which talks about one kind of contexts, e.g. return types, function parameter types, or default template arguments.

**Comment by AlisdairM** [comment]
`typename` for type aliases has been removed consistently with several recent PRs:
  #8371, #8376, and #8403

**Comment by eisenwave** [comment]
https://github.com/cplusplus/nbballot/issues/706 made a decision on this.

**Comment by jensmaurer** [comment]
> https://github.com/cplusplus/nbballot/issues/706 made a decision on this.

No, this one just made a decision on the very specific case `using T = typename U:whatever`, not on any other use of `typename`.  In contrast, P2150 as a whole seeks to eliminate numerous classes of `typename`.
